sgkit/io/bgen/bgen_reader.py

```diff
@@ -0,0 +1,618 @@
+"""BGEN reader implementation (using bgen_reader)"""
+import logging
+import tempfile
+from pathlib import Path
+from typing import (
+    Any,
+    Dict,
+    Hashable,
+    List,
+    Mapping,
+    MutableMapping,
+    Optional,
+    Tuple,
+    Union,
+)
+
+import dask
+import dask.array as da
+import dask.dataframe as dd
+import numpy as np
+import pandas as pd
+import xarray as xr
+import zarr
+from cbgen import bgen_file, bgen_metafile
+from rechunker import api as rechunker_api
+from xarray import Dataset
+
+from sgkit import create_genotype_dosage_dataset
+from sgkit.typing import ArrayLike, DType, PathType
+
+from ..utils import dataframe_to_dict, encode_contigs
+
+logger = logging.getLogger(__name__)
+
+GT_DATA_VARS = [
+    "call_genotype_probability",
+    "call_genotype_probability_mask",
+    "call_dosage",
+    "call_dosage_mask",
+]
+
+METAFILE_FIELDS = [
+    ("id", "S"),
+    ("rsid", "S"),
+    ("chrom", "S"),
+    ("pos", "int32"),
+    ("a1", "S"),
+    ("a2", "S"),
+    ("offset", "int64"),
+]
+METAFILE_DTYPE = dict(METAFILE_FIELDS)
+
+
+class BgenReader:
+
+    name = "bgen_reader"
+
+    def __init__(
+        self,
+        path: PathType,
+        metafile_path: Optional[PathType] = None,
+        dtype: DType = "float32",
+    ) -> None:
+        self.path = Path(path)
+        self.metafile_path = (
+            Path(metafile_path) if metafile_path else path.with_suffix(".metafile")  # type: ignore[union-attr]
+        )
+
+        with bgen_file(self.path) as bgen:
+            self.n_variants = bgen.nvariants
+            self.n_samples = bgen.nsamples
+
+            if not self.metafile_path.exists():
+                logger.info(
+                    f"Generating BGEN metafile for '{self.path}' (this may take a while)"
+                )
+                bgen.create_metafile(self.metafile_path, verbose=False)
+                logger.info("BGEN metafile generation complete")
```

Comment:
Diff'ing the timestamps in the logs is what I've been doing rather than adding more code.  Is there some precedent/good reason to start adding timers?

Reply:
Yea, sometimes this becomes useful if you have tons of logs in production (or even tests) job coming from different places (say in debug mode), and you don't need to search for two corresponding log entries.

Reply:
Hm well this code won't run on distributed workers (for the same bgen file) and I doubt that it ever will but 🤷‍♂️: https://github.com/pystatgen/sgkit/pull/314/commits/586aa9d167d71dffb5049edb8574957ea2da202d#diff-845c788aeac5e31bc9b2635b6fd903f0R82